optimism/material/HyperViscoelastic.py

```diff
@@ -0,0 +1,216 @@
+    assert 'non equilibrium shear modulus' in properties.keys()
+    assert 'relaxation time' in properties.keys()
+
+    print('Equilibrium properties')
+    print('  Bulk modulus    = %s' % properties['equilibrium bulk modulus'])
+    print('  Shear modulus   = %s' % properties['equilibrium shear modulus'])
+    print('Prony branch properties')
+    print('  Shear modulus   = %s' % properties['non equilibrium shear modulus'])
+    print('  Relaxation time = %s' % properties['relaxation time'])
+    # this is dirty, fuck jax (can't use an int from a jax numpy array or else jit tries to trace that)
+    # global NUM_PRONY_TERMS
+    # NUM_PRONY_TERMS = properties['number of prony terms']
+
+    # first pack equilibrium properties
+    props = np.array([
+        properties['equilibrium bulk modulus'],
+        properties['equilibrium shear modulus'],
+        properties['non equilibrium shear modulus'],
+        properties['relaxation time']
+    ])
+
+    # props = np.hstack((props, properties['number of prony terms']))
+
+    # for n in range(1, properties['number of prony terms'] + 1):
+    #     print('Prony branch %s properties' % n)
+    #     print('  Shear modulus   = %s' % properties['non equilibrium shear modulus %s' % n])
+    #     print('  Relaxation time = %s' % properties['relaxation time %s' % n])
+    #     props = np.hstack(
+    #         (props, np.array([properties['non equilibrium shear modulus %s' % n],
+    #                           properties['relaxation time %s' % n]])))
+
+    
+
+    return props
+
+def _energy_density(dispGrad, state, dt, props):
+    W_eq  = _eq_strain_energy(dispGrad, props)
+    W_neq = _neq_strain_energy(dispGrad, state, dt, props)
+    return W_eq + W_neq
+
+# TODO generalize to arbitrary strain energy density
+def _eq_strain_energy(dispGrad, props):
+    K, G = props[PROPS_K_eq], props[PROPS_G_eq]
+    F = dispGrad + np.eye(3)
+    J = np.linalg.det(F)
+    J23 = np.power(J, -2.0 / 3.0)
+    I1Bar = J23 * np.tensordot(F,F)
+    Wvol = 0.5 * K * (0.5 * J**2 - 0.5 - np.log(J))
+    Wdev = 0.5 * G * (I1Bar - 3.0)
+    return Wdev + Wvol
+
+def _neq_strain_energy(dispGrad, stateOld, dt, props):
+    I = np.identity(3)
+    F = dispGrad + I
+    state_new = _compute_state_new(dispGrad, stateOld, dt, props)
+    # Fvs = state_new.reshape((NUM_PRONY_TERMS, 3, 3))
+    Fv_new = state_new.reshape((3, 3))
+
+    G_neq = props[PROPS_G_neq]
+    tau   = props[PROPS_TAU]
+    eta   = G_neq * tau
+
+    Fe = F @ np.linalg.inv(Fv_new)
+    Ee = 0.5 * TensorMath.mtk_log_sqrt(Fe.T @ Fe)
+    Me = 2. * G_neq * Ee
+    M_bar = TensorMath.norm_of_deviator_squared(Me)
+    gamma_dot = M_bar / eta
+    # visco_energy = (dt / (G_neq * tau)) * M_bar**2
+    visco_energy = 0.5 * dt * eta * gamma_dot**2
+
+    W_neq = G_neq * TensorMath.norm_of_deviator_squared(Ee) + visco_energy
+    # def vmap_body(n, Fv):
+    #     G_neq = props[PROPS_G_neq + 2 * n]
+    #     # tau   = props[PROPS_TAU   + 2 * n]
```

Comment:
FWIW, If we end up implementing multiple Prony terms, I'd suggest trying with a `jax.lax.fori` loop, or `jax.lax.scan`. I think it might be more straightforward than `vmap`.